### ISSUE TITLE
CapsNet: bad initialization

### ISSUE THREAD

**Issue by arodiss** [ageron/handson-ml#164]
Hi @ageron 

First of all I'd like to thank you for this awesome intros, and also for explanatory video for CapsNet
I followed your instruction there and than hacked model around.

It looks like your initialization for transformation matrices between primary capsules and digit capsules is very unfortunate. It typically takes ~50 batches (I use batch size 16) to start changing losses in visible range. This is a typical sign of bad initialization.
After this initial delay training takes on quite rapidly, so it does not look like bad LR
When I change stddev of initial weights from 0.01 to 0.1, models starts visible learning right away

So IMHO it makes sense to change value in notebook as well

**Comment by ageron** [comment]
Hi Gleb,

Thanks for your feedback: that's a great point, I didn't explore the hyperparameters, I just tried to reproduce the results in the paper (and they also explain in the paper that they didn't explore the space much either, it's mostly a proof-of-concept, not a tuned solution).

FYI, Thibault Neveu also suggested using `tf.einsum()` rather than `tf.tile()`. Training is much faster that way. Here's his message:

> I implement capsNet using broadcasting with matrix multiplication without `tf.tile()` by using `tf.einsum()`:
> `u_hat = tf.einsum('abdc,iabcf->iabdf', w_ij, input_layer)`
> It's more computationally efficient

Thanks again, I'll update the notebook as soon as I have a bit of time (or don't hesitate to submit a Pull Request, that'd be awesome).

Thanks again!
Aurélien

**Comment by arodiss** [comment]
@ageron
Here it is: #176

**Comment by neerajtiwari360** [comment]
Hi @ageron 

I have a dought in line 14 of extra_capsnets.ipynb


In [14]:
conv1 = tf.layers.conv2d(X, name="conv1", **conv1_params)
conv2 = tf.layers.conv2d(conv1, name="conv2", **conv2_params)
Note: since we used a kernel size of 9 and no padding (for some reason, that's what "valid" means), the image shrunk by 9-1=8 pixels after each convolutional layer (28×28 to 20×20, then 20×20 to 12×12), and since we used a stride of 2 in the second convolutional layer, the image size was divided by 2. This is how we end up with 6×6 feature maps.

Next, we reshape the output to get a bunch of 8D vectors representing the outputs of the primary capsules. The output of conv2 is an array containing 32×8=256 feature maps for each instance, where each feature map is 6×6. So the shape of this output is (batch size, 6, 6, 256). We want to chop the 256 into 32 vectors of 8 dimensions each. We could do this by reshaping to (batch size, 6, 6, 32, 8). However, since this first capsule layer will be fully connected to the next capsule layer, we can simply flatten the 6×6 grids. This means we just need to reshape to (batch size, 6×6×32, 8).


My Question: here in last line as you mention  that shape will be (batch size, 6×6×32, 8) but after reading paper Dynamic Routing Between Capsules it seems like the shape will be (batch size, 6×6×8, 32)

**Comment by ageron** [comment]
Hi @neerajtiwari360 ,
Thanks for your question. I double-checked the paper, and I cannot find any mention of a (batch size, 6x6x8, 32) shape.  Instead, I found the following text:

> In total PrimaryCapsules has [32 × 6 × 6] capsule outputs (each output is an 8D vector) and each capsule in the [6 × 6] grid is sharing their weights with each other.

Hope this helps,
Aurélien
